examples/vision/edsr.py

```diff
@@ -0,0 +1,276 @@
+
+
+"""
+# Model Building(Baseline Model)
+1. 16 residual blocks
+2. filters=64
+3. No batchNormalization
+"""
+
+"""
+<img src = "https://miro.medium.com/max/1400/1*a1tQ4grku_ugTTe39IJzjA.png" width="500">
+"""
+
+
+def ResBlock(inputs):
+    x = tf.keras.layers.Conv2D(64, 3, padding="same", activation="relu")(inputs)
+    x = tf.keras.layers.Conv2D(64, 3, padding="same")(x)
+    x = tf.keras.layers.Add()([inputs, x])
+    return x
+
+
+def Upsampling(inputs, factor=2, **kwargs):
+    x = tf.keras.layers.Conv2D(64 * (factor**2), 3, padding="same", **kwargs)(inputs)
+    x = tf.keras.layers.Lambda(shuffle_pixels(scale=factor))(x)
+    x = tf.keras.layers.Conv2D(64 * (factor**2), 3, padding="same", **kwargs)(x)
+    x = tf.keras.layers.Lambda(shuffle_pixels(scale=factor))(x)
+    return x
+
+
+def EDSR():
+    input_layer = tf.keras.layers.Input(shape=(None, None, 3))
+    x = tf.keras.layers.Lambda(normalize)(input_layer)
+    x = x_new = tf.keras.layers.Conv2D(64, 3, padding="same")(x)
+
+    # 16 residual blocks
+    for _ in range(16):
+        x_new = ResBlock(x_new)
+
+    x_new = tf.keras.layers.Conv2D(64, 3, padding="same")(x_new)
+    x = tf.keras.layers.Add()([x, x_new])
+
+    x = Upsampling(x)
+    x = tf.keras.layers.Conv2D(3, 3, padding="same")(x)
+    output_layer = tf.keras.layers.Lambda(denormalize)(x)
+    return tf.keras.models.Model(input_layer, output_layer)
+
+
+edsr = EDSR()
+edsr.summary()
+
+"""
+# Training
+EDSR can take a long time to train, in the code below, but we will train for 5 epochs and
+20 steps for checking if the code is working alright. In practice, a larger steps value
+(over 200000) is required to get decent results.
+
+
+paper Suggest that learning rate of adam optimizer will be halfed after 2 × 10^5 steps.
+But as we are training for only 5 epochs we will change it to 50 steps.
+
+"""
+
+loss_fn = tf.keras.losses.MeanAbsoluteError()
+optim_edsr = tf.keras.optimizers.Adam(
+    learning_rate=PiecewiseConstantDecay(boundaries=[50], values=[1e-4, 5e-5])
+)
+
+
+@tf.function
+def train_step(ds_low, ds_high):
+    with tf.GradientTape() as EDSR:
+
+        ds_low = tf.cast(ds_low, tf.float32)
+        ds_high = tf.cast(ds_high, tf.float32)
+
+        sr = edsr(ds_low, training=True)
+        loss_value = loss_fn(ds_high, sr)
+
+        # Calculating PSNR value
+        psnr_value = PSNR(ds_high, sr)
+
+    gradients = EDSR.gradient(loss_value, edsr.trainable_variables)
+    optim_edsr.apply_gradients(zip(gradients, edsr.trainable_variables))
+
+    return loss_value, psnr_value
+
+
+"""
+##### Training for 5*20 = 100 steps
+"""
+
+for epoch in range(5):
+
+    for lr, hr in ds.take(20):
+        loss_value, psnr_value = train_step(lr, hr)
+
+    print(
+        f"Epochs : {epoch}   ||   Loss : {loss_value:.5f}   ||   PSNR : {psnr_value:.5f}"
+    )
+
+"""
+# Saving Model
+"""
+
+"""!mkdir model
+edsr.save('saved_model/',save_format='tf')
+edsr.save("saved_model/model.h5")"""
+
+"""
+# Testing Model
```

Comment:
Every section should have enough text explanations to make for an easy-to-follow (perhaps even compelling) read.